examples/google-cloud-build/README.md

```diff
@@ -0,0 +1,81 @@
+# Use [Google Cloud Build](https://cloud.google.com/cloud-build) with Jib
+
+This example shows how to use [Google Cloud Build](https://cloud.google.com/cloud-build) and Jib to build your Java project into a container image. 
+
+## Prerequisites
+
+Make sure you enable the Cloud Build API and have the Cloud SDK installed. The `gcloud` binary should be on your `PATH`.
+
+See the [Cloud Build Quickstart](https://cloud.google.com/cloud-build/docs/quickstart-docker#before-you-begin) for instructions.
+
+## Build a Maven project
+
+To build your Maven project into a container image, use [`cloudbuild-jib-maven.yaml`](cloudbuild-jib-maven.yaml) as the Cloud Build build configuration file:
+
+```bash
+gcloud builds submit --config cloudbuild-jib-maven.yaml <Maven project directory>
+```
+
+Make sure to change the `_IMAGE_NAME` substitution to the name of the image you wish to build.
+
+## Build a Gradle project
+
+To build your Gradle project into a container image, first add Jib to your `build.gradle`:
+
+```groovy
+plugins {
+  id 'com.google.cloud.tools.jib' version '1.0.0'
+}
+```
+
+Then, use [`cloudbuild-jib-gradle.yaml`](cloudbuild-jib-gradle.yaml) as the Cloud Build build configuration file:
+
+```bash
+gcloud builds submit --config cloudbuild-jib-gradle.yaml <Gradle project directory>
+```
+
+Make sure to change the `_IMAGE_NAME` substitution to the name of the image you wish to build.
+
+### Build a Gradle project without applying the Jib plugin
+
+You can also build any Gradle project without having to add Jib to the `build.gradle`. First, copy [`auto-jib.gradle`](auto-jib.gradle) to your Gradle project directory. Then, uncomment the `init-script` arg in [`cloudbuild-jib-gradle.yaml`](cloudbuild-jib-gradle-auto.yaml) and submit the build with:
+
+```bash
+gcloud builds submit --config cloudbuild-jib-gradle.yaml <Gradle project directory>
+``` 
+
+## HelloWorld example
+
+Try building the provided [helloworld-sample](helloworld-sample) Java project on Google Cloud Build:
+
+### Maven
+
+```bash
+gcloud builds submit --config cloudbuild-jib-maven.yaml helloworld-sample
+```
+
+### Gradle
+
+Uncomment the `init-script` arg in [`cloudbuild-jib-gradle.yaml`](cloudbuild-jib-gradle.yaml) and submit the build with:
+
+```bash
+gcloud builds submit --config cloudbuild-jib-gradle.yaml helloworld-sample
+```
+
+## Speeding up your build
+
+By default, Google Cloud Build does not cache anything across builds. However, you can [use Google Cloud Storage to cache files](https://cloud.google.com/cloud-build/docs/speeding-up-builds) across builds to speed up subsequent builds.
```

Comment:
Is this strictly true? I remember vaguely that GCS sometimes didn't speed anything up?

Reply:
GCS isn't guaranteed to speed anything up, it depends a lot on your use case, what you write back to GCS at the end of a build and how much you can reuse on the next build (and file sizes/counts, since writing 1000000 1B files to GCS is tons slower than one 1MB file)

Similarly, how you upload stuff to GCS and how you read it back can matter a lot to how concurrency-safe and incremental it is.

Reply:
Yea, probably not strictly true, but redownloading lots of Maven dependencies each time seems slower than downloading the cache tar and extracting. Based on my runs for the helloworld example, the caching brings the total build time down from 40+ seconds to 30+ seconds. As for concurrency, since single object writes are transactional, concurrent writes would not create corrupt caches, and for our use case, stale writes should be fine since it's just a cache. Incrementality would not be present though since the caches are bundled up as single objects.